File: app/src/main/java/dev/adamsjack/beanshelf/ui/CropEditor.kt
package dev.adamsjack.beanshelf.ui

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.ui.input.pointer.PointerInputScope
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import dev.adamsjack.beanshelf.data.PhotoStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.util.UUID
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/**
 * Full-screen manual crop: drag inside the frame to move it, drag a corner to
 * resize. Crop writes a NEW file (old one is deleted) so every cache and
 * composable refreshes off the path change. PNG stays PNG (cutout alpha kept).
 */
@Composable
fun CropDialog(path: String, onDone: (String?) -> Unit) {
    val scope = rememberCoroutineScope()
    var rotation by remember { mutableStateOf(0) } // 0/90/180/270, clockwise
    val loaded by produceState<Bitmap?>(initialValue = null, path) {
        value = withContext(Dispatchers.IO) { BitmapFactory.decodeFile(path) }
    }
    // Camera EXIF is sometimes wrong — the Rotate button makes straightening deterministic.
    val bitmap = remember(loaded, rotation) {
        val src = loaded
        if (src == null || rotation == 0) src
        else Bitmap.createBitmap(
            src, 0, 0, src.width, src.height,
            android.graphics.Matrix().apply { postRotate(rotation.toFloat()) }, true,
        )
    }

    Dialog(
        onDismissRequest = { onDone(null) },
        properties = DialogProperties(usePlatformDefaultWidth = false),
    ) {
        Box(Modifier.fillMaxSize().background(Color.Black.copy(alpha = 0.96f))) {
            val bmp = bitmap ?: return@Box
            var container by remember { mutableStateOf(IntSize.Zero) }
            // Fit-rect of the displayed image inside the container.
            val imageRect = remember(container, bmp) {
                if (container == IntSize.Zero) Rect.Zero else {
                    val scale = min(
                        container.width / bmp.width.toFloat(),
                        container.height / bmp.height.toFloat(),
                    )
                    val w = bmp.width * scale
                    val h = bmp.height * scale
                    val l = (container.width - w) / 2f
                    val t = (container.height - h) / 2f
                    Rect(l, t, l + w, t + h)
                }
            }
            var crop by remember(imageRect) { mutableStateOf(imageRect) }
            val touchRadius = with(LocalDensity.current) { 36.dp.toPx() }
            val minSide = with(LocalDensity.current) { 56.dp.toPx() }

            Image(
                bitmap = bmp.asImageBitmap(),
                contentDescription = "Crop photo",
                contentScale = ContentScale.Fit,
                modifier = Modifier
                    .fillMaxSize()
                    .padding(bottom = 96.dp)
                    .onSizeChanged { container = it },
            )

            if (imageRect != Rect.Zero) {
                Canvas(
                    Modifier
                        .fillMaxSize()
                        .padding(bottom = 96.dp)
                        .pointerInput(imageRect) {
                            var mode = ""
                            detectDragGesturesLocal(
                                onStart = { pos ->
                                    val corners = mapOf(
                                        "tl" to Offset(crop.left, crop.top),
                                        "tr" to Offset(crop.right, crop.top),
                                        "bl" to Offset(crop.left, crop.bottom),
                                        "br" to Offset(crop.right, crop.bottom),
                                    )
                                    mode = corners.entries
                                        .firstOrNull { (it.value - pos).getDistance() < touchRadius }
                                        ?.key
                                        ?: if (crop.contains(pos)) "move" else ""
                                },
                                onDrag = { drag ->
                                    val r = crop
                                    crop = when (mode) {
                                        "move" -> {
                                            val dx = drag.x.coerceIn(imageRect.left - r.left, imageRect.right - r.right)
                                            val dy = drag.y.coerceIn(imageRect.top - r.top, imageRect.bottom - r.bottom)
                                            r.translate(dx, dy)
                                        }
                                        "tl" -> Rect(
                                            (r.left + drag.x).coerceIn(imageRect.left, r.right - minSide),
                                            (r.top + drag.y).coerceIn(imageRect.top, r.bottom - minSide),
                                            r.right, r.bottom,
                                        )
                                        "tr" -> Rect(
                                            r.left,
                                            (r.top + drag.y).coerceIn(imageRect.top, r.bottom - minSide),
                                            (r.right + drag.x).coerceIn(r.left + minSide, imageRect.right),
                                            r.bottom,
                                        )
                                        "bl" -> Rect(
                                            (r.left + drag.x).coerceIn(imageRect.left, r.right - minSide),
                                            r.top, r.right,
                                            (r.bottom + drag.y).coerceIn(r.top + minSide, imageRect.bottom),
                                        )
                                        "br" -> Rect(
                                            r.left, r.top,
                                            (r.right + drag.x).coerceIn(r.left + minSide, imageRect.right),
                                            (r.bottom + drag.y).coerceIn(r.top + minSide, imageRect.bottom),
                                        )
                                        else -> r
                                    }
                                },
                            )
                        },
                ) {
                    // dim everything outside the crop frame
                    drawRect(Color.Black.copy(alpha = 0.55f), Offset.Zero, Size(size.width, crop.top))
                    drawRect(Color.Black.copy(alpha = 0.55f), Offset(0f, crop.bottom), Size(size.width, size.height - crop.bottom))
                    drawRect(Color.Black.copy(alpha = 0.55f), Offset(0f, crop.top), Size(crop.left, crop.height))
                    drawRect(Color.Black.copy(alpha = 0.55f), Offset(crop.right, crop.top), Size(size.width - crop.right, crop.height))
                    drawRect(Crema, crop.topLeft, crop.size, style = Stroke(2.dp.toPx()))
                    listOf(
                        Offset(crop.left, crop.top), Offset(crop.right, crop.top),
                        Offset(crop.left, crop.bottom), Offset(crop.right, crop.bottom),
                    ).forEach { drawCircle(Crema, radius = 7.dp.toPx(), center = it) }
                }
            }

            Row(
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .fillMaxWidth()
                    .navigationBarsPadding()
                    .padding(horizontal = 20.dp, vertical = 16.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                TextButton(onClick = { onDone(null) }) { Text("Cancel", color = Dim) }
                TextButton(onClick = { rotation = (rotation + 90) % 360 }) {
                    Text("Rotate ⟳", color = Crema)
                }
                TextButton(onClick = { crop = imageRect; rotation = 0 }, modifier = Modifier.padding(start = 4.dp)) {
                    Text("Reset", color = Parchment)
                }
                Box(Modifier.weight(1f))
                Button(
                    onClick = {
                        scope.launch {
                            val newPath = withContext(Dispatchers.IO) {
                                applyCrop(bmp, path, crop, imageRect, force = rotation != 0)
                            }
                            onDone(newPath)
                        }
                    },
                    colors = ButtonDefaults.buttonColors(containerColor = Crema, contentColor = Roast),
                ) { Text("Save") }
            }
        }
    }
}

/** Maps the display-space frame back to bitmap pixels and writes a new file. */
private fun applyCrop(bmp: Bitmap, oldPath: String, crop: Rect, imageRect: Rect, force: Boolean = false): String? =
    runCatching {
        val scale = bmp.width / imageRect.width
        val l = ((crop.left - imageRect.left) * scale).toInt().coerceIn(0, bmp.width - 1)
        val t = ((crop.top - imageRect.top) * scale).toInt().coerceIn(0, bmp.height - 1)
        val w = (crop.width * scale).toInt().coerceIn(1, bmp.width - l)
        val h = (crop.height * scale).toInt().coerceIn(1, bmp.height - t)
        // No-op crop (and no rotation) → keep the original file.
        if (!force && l == 0 && t == 0 && abs(w - bmp.width) < 4 && abs(h - bmp.height) < 4) return@runCatching oldPath

        val cropped = Bitmap.createBitmap(bmp, l, t, w, h)
        val old = File(oldPath)
        val isPng = oldPath.endsWith(".png")
        val newFile = File(old.parentFile, "${UUID.randomUUID()}${if (isPng) ".png" else ".jpg"}")
        newFile.outputStream().use {
            cropped.compress(
                if (isPng) Bitmap.CompressFormat.PNG else Bitmap.CompressFormat.JPEG,
                if (isPng) 100 else 88,
                it,
            )
        }
        old.delete()
        newFile.absolutePath
    }.getOrNull()

/** Minimal drag detector with a start callback. */
private suspend fun PointerInputScope.detectDragGesturesLocal(
    onStart: (Offset) -> Unit,
    onDrag: (Offset) -> Unit,
) {
    detectDragGestures(
        onDragStart = { onStart(it) },
        onDrag = { change, amount ->
            change.consume()
            onDrag(amount)
        },
    )
}

File: ios/Beanshelf/Views/CropEditor.swift
import SwiftUI
import UIKit

/// Full-screen manual crop: drag inside the frame to move it, drag a corner to
/// resize, Rotate ⟳ turns in 90° steps (camera EXIF is sometimes wrong — the
/// button makes straightening deterministic). Crop writes a NEW file (old one
/// is deleted) so every cache and view refreshes off the filename change.
/// PNG stays PNG (cutout alpha kept).
struct CropEditorView: View {
    let file: String
    let onDone: (String?) -> Void

    @State private var loaded: UIImage?
    @State private var rotation = 0 // 0/90/180/270, clockwise
    @State private var crop: CGRect = .zero
    @State private var imageRect: CGRect = .zero
    @State private var dragMode: String?
    @State private var lastTranslation: CGSize = .zero
    @State private var saving = false

    private let touchRadius: CGFloat = 36
    private let minSide: CGFloat = 56

    private var bitmap: UIImage? {
        guard let loaded else { return nil }
        return rotation == 0 ? loaded : loaded.rotatedClockwise(steps: rotation / 90)
    }

    var body: some View {
        ZStack {
            Color.black.opacity(0.96).ignoresSafeArea()
            VStack(spacing: 0) {
                GeometryReader { geo in
                    ZStack {
                        if let bitmap {
                            Image(uiImage: bitmap)
                                .resizable()
                                .scaledToFit()
                                .frame(width: geo.size.width, height: geo.size.height)
                        }
                        Canvas { context, size in
                            guard imageRect != .zero else { return }
                            let dim = Color.black.opacity(0.55)
                            context.fill(Path(CGRect(x: 0, y: 0, width: size.width, height: crop.minY)), with: .color(dim))
                            context.fill(Path(CGRect(x: 0, y: crop.maxY, width: size.width, height: size.height - crop.maxY)), with: .color(dim))
                            context.fill(Path(CGRect(x: 0, y: crop.minY, width: crop.minX, height: crop.height)), with: .color(dim))
                            context.fill(Path(CGRect(x: crop.maxX, y: crop.minY, width: size.width - crop.maxX, height: crop.height)), with: .color(dim))
                            context.stroke(Path(crop), with: .color(Palette.crema), lineWidth: 2)
                            for corner in [
                                CGPoint(x: crop.minX, y: crop.minY), CGPoint(x: crop.maxX, y: crop.minY),
                                CGPoint(x: crop.minX, y: crop.maxY), CGPoint(x: crop.maxX, y: crop.maxY),
                            ] {
                                context.fill(
                                    Path(ellipseIn: CGRect(x: corner.x - 7, y: corner.y - 7, width: 14, height: 14)),
                                    with: .color(Palette.crema)
                                )
                            }
                        }
                        .contentShape(Rectangle())
                        .gesture(dragGesture)
                    }
                    .onAppear { recomputeImageRect(container: geo.size) }
                    .onChange(of: geo.size) { _, size in recomputeImageRect(container: size) }
                    .onChange(of: rotation) { _, _ in recomputeImageRect(container: geo.size) }
                    .onChange(of: loaded == nil) { _, _ in recomputeImageRect(container: geo.size) }
                }

                HStack(spacing: 4) {
                    Button("Cancel") { onDone(nil) }
                        .foregroundStyle(Palette.dim)
                    Button("Rotate ⟳") { rotation = (rotation + 90) % 360 }
                        .foregroundStyle(Palette.crema)
                    Button("Reset") { crop = imageRect; rotation = 0 }
                        .foregroundStyle(Palette.parchment)
                        .padding(.leading, 4)
                    Spacer()
                    Button {
                        guard !saving, let bitmap else { return }
                        saving = true
                        let cropRect = crop
                        let fitRect = imageRect
                        let force = rotation != 0
                        Task {
                            let newFile = await applyCrop(bitmap: bitmap, cropRect: cropRect, fitRect: fitRect, force: force)
                            onDone(newFile)
                        }
                    } label: {
                        Text("Save")
                            .font(Type.labelLarge)
                            .foregroundStyle(Palette.roast)
                            .padding(.horizontal, 22)
                            .padding(.vertical, 10)
                            .background(Palette.crema, in: Capsule())
                    }
                }
                .font(Type.labelLarge)
                .padding(.horizontal, 20)
                .padding(.vertical, 16)
            }
        }
        .task {
            guard let url = PhotoStore.url(for: file) else { return }
            loaded = await Task.detached { UIImage(contentsOfFile: url.path) }.value
        }
    }

    /// Fit-rect of the displayed image inside the container.
    private func recomputeImageRect(container: CGSize) {
        guard let bitmap, container != .zero else { return }
        let scale = min(container.width / bitmap.size.width, container.height / bitmap.size.height)
        let w = bitmap.size.width * scale
        let h = bitmap.size.height * scale
        imageRect = CGRect(x: (container.width - w) / 2, y: (container.height - h) / 2, width: w, height: h)
        crop = imageRect
    }

    private var dragGesture: some Gesture {
        DragGesture(minimumDistance: 0)
            .onChanged { value in
                if dragMode == nil {
                    lastTranslation = .zero
                    let pos = value.startLocation
                    let corners = [
                        "tl": CGPoint(x: crop.minX, y: crop.minY),
                        "tr": CGPoint(x: crop.maxX, y: crop.minY),
                        "bl": CGPoint(x: crop.minX, y: crop.maxY),
                        "br": CGPoint(x: crop.maxX, y: crop.maxY),
                    ]
                    dragMode = corners.first { hypot($0.value.x - pos.x, $0.value.y - pos.y) < touchRadius }?.key
                        ?? (crop.contains(pos) ? "move" : "")
                }
                let dx = value.translation.width - lastTranslation.width
                let dy = value.translation.height - lastTranslation.height
                lastTranslation = value.translation
                applyDrag(dx: dx, dy: dy)
            }
            .onEnded { _ in
                dragMode = nil
                lastTranslation = .zero
            }
    }

    private func applyDrag(dx: CGFloat, dy: CGFloat) {
        let r = crop
        switch dragMode {
        case "move":
            let cx = min(max(dx, imageRect.minX - r.minX), imageRect.maxX - r.maxX)
            let cy = min(max(dy, imageRect.minY - r.minY), imageRect.maxY - r.maxY)
            crop = r.offsetBy(dx: cx, dy: cy)
        case "tl":
            let nl = min(max(r.minX + dx, imageRect.minX), r.maxX - minSide)
            let nt = min(max(r.minY + dy, imageRect.minY), r.maxY - minSide)
            crop = CGRect(x: nl, y: nt, width: r.maxX - nl, height: r.maxY - nt)
        case "tr":
            let nr = min(max(r.maxX + dx, r.minX + minSide), imageRect.maxX)
            let nt = min(max(r.minY + dy, imageRect.minY), r.maxY - minSide)
            crop = CGRect(x: r.minX, y: nt, width: nr - r.minX, height: r.maxY - nt)
        case "bl":
            let nl = min(max(r.minX + dx, imageRect.minX), r.maxX - minSide)
            let nb = min(max(r.maxY + dy, r.minY + minSide), imageRect.maxY)
            crop = CGRect(x: nl, y: r.minY, width: r.maxX - nl, height: nb - r.minY)
        case "br":
            let nr = min(max(r.maxX + dx, r.minX + minSide), imageRect.maxX)
            let nb = min(max(r.maxY + dy, r.minY + minSide), imageRect.maxY)
            crop = CGRect(x: r.minX, y: r.minY, width: nr - r.minX, height: nb - r.minY)
        default:
            break
        }
    }

    /// Maps the display-space frame back to bitmap pixels and writes a new file.
    private func applyCrop(bitmap: UIImage, cropRect: CGRect, fitRect: CGRect, force: Bool) async -> String? {
        await Task.detached(priority: .userInitiated) { () -> String? in
            guard let cg = bitmap.cgImage else { return nil }
            let scale = CGFloat(cg.width) / fitRect.width
            let l = Int(((cropRect.minX - fitRect.minX) * scale).rounded())
            let t = Int(((cropRect.minY - fitRect.minY) * scale).rounded())
            let w = Int((cropRect.width * scale).rounded())
            let h = Int((cropRect.height * scale).rounded())
            let cl = min(max(l, 0), cg.width - 1)
            let ct = min(max(t, 0), cg.height - 1)
            let cw = min(max(w, 1), cg.width - cl)
            let ch = min(max(h, 1), cg.height - ct)
            // No-op crop (and no rotation) → keep the original file.
            if !force && cl == 0 && ct == 0 && abs(cw - cg.width) < 4 && abs(ch - cg.height) < 4 {
                return file
            }
            guard let cropped = cg.cropping(to: CGRect(x: cl, y: ct, width: cw, height: ch)) else { return nil }
            let isPng = file.hasSuffix(".png")
            let out = UIImage(cgImage: cropped)
            guard let data = isPng ? out.pngData() : out.jpegData(compressionQuality: 0.88) else { return nil }
            let newFile = "\(UUID().uuidString).\(isPng ? "png" : "jpg")"
            guard let newURL = PhotoStore.url(for: newFile) else { return nil }
            do {
                try data.write(to: newURL, options: .atomic)
                PhotoStore.delete(file)
                return newFile
            } catch {
                return nil
            }
        }.value
    }
}

extension UIImage {
    /// Bakes a clockwise rotation of 90° × steps into the pixels.
    func rotatedClockwise(steps: Int) -> UIImage {
        let n = ((steps % 4) + 4) % 4
        guard n != 0, let cg = cgImage else { return self }
        let orientation: UIImage.Orientation = [.up, .right, .down, .left][n]
        let oriented = UIImage(cgImage: cg, scale: 1, orientation: orientation)
        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        return UIGraphicsImageRenderer(size: oriented.size, format: format).image { _ in
            oriented.draw(in: CGRect(origin: .zero, size: oriented.size))
        }
    }
}
